diff --git a/toys/pending/chsh.c b/toys/pending/chsh.c
index 8d3390ee..512137f1 100644
--- a/toys/pending/chsh.c
+++ b/toys/pending/chsh.c
@@ -71,7 +71,7 @@ void chsh_main()
 
   // Update /etc/passwd
   passwd_info->pw_shell = line;
-  if (-1 == update_password("/etc/passwd", user, NULL)) perror_exit("Failed to remove passwd entry");
+  if (-1 == update_password("/etc/passwd", user, NULL, 1)) perror_exit("Failed to remove passwd entry");
   file = xfopen("/etc/passwd", "a");
   if (putpwent(passwd_info, file)) perror_exit("putwent");
 }
diff --git a/toys/pending/groupadd.c b/toys/pending/groupadd.c
index 642b4a08..eab83288 100644
--- a/toys/pending/groupadd.c
+++ b/toys/pending/groupadd.c
@@ -50,10 +50,10 @@ static void new_group()
   }
 
   entry = xmprintf("%s:%s:%ld:", *toys.optargs, "x", TT.gid);
-  update_password(GROUP_PATH, *toys.optargs, entry);
+  update_password(GROUP_PATH, *toys.optargs, entry, 1);
   free(entry);
   entry = xmprintf("%s:%s::", *toys.optargs, "!");
-  update_password(SECURE_GROUP_PATH, *toys.optargs, entry);
+  update_password(SECURE_GROUP_PATH, *toys.optargs, entry, 1);
   free(entry);
 }
 
@@ -86,8 +86,8 @@ void groupadd_main(void)
       entry = xrealloc(entry, strlen(entry) + strlen(*toys.optargs) + 1);
       strcat(entry, *toys.optargs);
     }
-    update_password(GROUP_PATH, grp->gr_name, entry);
-    update_password(SECURE_GROUP_PATH, grp->gr_name, entry);
+    update_password(GROUP_PATH, grp->gr_name, entry, 1);
+    update_password(SECURE_GROUP_PATH, grp->gr_name, entry, 1);
     free(entry);
   } else {    //new group to be created
     char *s = *toys.optargs;
diff --git a/toys/pending/groupdel.c b/toys/pending/groupdel.c
index 483ac59c..bac755b1 100644
--- a/toys/pending/groupdel.c
+++ b/toys/pending/groupdel.c
@@ -55,7 +55,7 @@ void groupdel_main(void)
     if (CFG_TOYBOX_FREE) endpwent();
   }
 
-  update_password("/etc/group", grp->gr_name, entry);
-  update_password("/etc/gshadow", grp->gr_name, entry);
+  update_password("/etc/group", grp->gr_name, entry, 1);
+  update_password("/etc/gshadow", grp->gr_name, entry ,1);
   if (CFG_TOYBOX_FREE) free(entry);
 }
diff --git a/toys/pending/useradd.c b/toys/pending/useradd.c
index d0ad03ab..9f2ea1e4 100644
--- a/toys/pending/useradd.c
+++ b/toys/pending/useradd.c
@@ -119,7 +119,7 @@ void useradd_main(void)
   entry = xmprintf("%s:%s:%ld:%ld:%s:%s:%s", pwd.pw_name, pwd.pw_passwd,
       (long)pwd.pw_uid, (long)pwd.pw_gid, pwd.pw_gecos, pwd.pw_dir,
       pwd.pw_shell);
-  if (update_password("/etc/passwd", pwd.pw_name, entry)) error_exit("updating passwd file failed");
+  if (update_password("/etc/passwd", pwd.pw_name, entry, 1)) error_exit("updating passwd file failed");
   free(entry);
 
   if (toys.optflags & FLAG_S) 
@@ -127,7 +127,7 @@ void useradd_main(void)
       (unsigned)(time(NULL))/(24*60*60)); //passwd is not set initially
   else entry = xmprintf("%s:!!:%u:0:99999:7:::", pwd.pw_name, 
             (unsigned)(time(0))/(24*60*60)); //passwd is not set initially
-  update_password("/etc/shadow", pwd.pw_name, entry);
+  update_password("/etc/shadow", pwd.pw_name, entry, 1);
   free(entry);
 
   // create home dir & copy skel dir to home
diff --git a/toys/pending/userdel.c b/toys/pending/userdel.c
index 51ee67c9..f522cbf7 100644
--- a/toys/pending/userdel.c
+++ b/toys/pending/userdel.c
@@ -92,8 +92,8 @@ void userdel_main(void)
 {
   struct passwd *pwd = xgetpwnam(*toys.optargs);
 
-  update_password("/etc/passwd", pwd->pw_name, NULL);
-  update_password("/etc/shadow", pwd->pw_name, NULL);
+  update_password("/etc/passwd", pwd->pw_name, NULL ,1);
+  update_password("/etc/shadow", pwd->pw_name, NULL, 1);
 
   // delete the group named USER, and remove user from group.
   // could update_password() be used for this? 
