--- a/meson.build
+++ b/meson.build
@@ -102,6 +102,7 @@
   'src/entry_backend/pango.c',
   'src/entry_backend/harfbuzz.c',
   'src/fuzzy_match.c',
+  'src/fts/fts.c',
   'src/history.c',
   'src/input.c',
   'src/lock.c',
@@ -129,8 +130,6 @@
 cc = meson.get_compiler('c')
 librt = cc.find_library('rt', required: false)
 libm = cc.find_library('m', required: false)
-# On systems where libc doesn't provide fts (i.e. musl) we require libfts
-libfts = cc.find_library('fts', required: not cc.has_function('fts_read'))
 freetype = dependency('freetype2')
 harfbuzz = dependency('harfbuzz')
 cairo = dependency('cairo')
@@ -191,7 +190,7 @@
 executable(
   'tofi',
   files('src/main.c'), common_sources, wl_proto_src, wl_proto_headers,
-  dependencies: [librt, libm, libfts, freetype, harfbuzz, cairo, pangocairo, wayland_client, xkbcommon, glib, gio_unix],
+  dependencies: [librt, libm, freetype, harfbuzz, cairo, pangocairo, wayland_client, xkbcommon, glib, gio_unix],
   install: true
 )
 
